[MSW] Encode wxUniChar value as UTF-16 when initializing wxString

minoki opened the pull request:
On MSW/wchar_t build, when initializing wxString with a wxUniChar value, the upper 16 bits are silently dropped.
 (See [Trac #11827](http://trac.wxwidgets.org/ticket/11827) and [Trac #16375](http://trac.wxwidgets.org/ticket/16375))

```c++
wxUniChar ch(0x12345);
wxString s(ch); // => L"\u2345"
```

With this PR, the wxUniChar value is encoded as UTF-16 instead of being converted to a wrong value.

Note that, on MSW/wchar_t build, wxString will still behave as UTF-16 sequence in other aspects.

minoki committed:
Add a test case for constructing wxString with supplementary wxUniChar values

See #11827

minoki committed:
Better handling of supplementary wxUniChar values in some of wxString methods

On MSW, the Unicode code point is now properly encoded as UTF-16 when
assigned or appended to a wxString.

Closes #11827

vadz commented on include/wx/string.h:
```diff
@@ -1824,7 +1824,11 @@ class WXDLLIMPEXP_BASE wxString
         m_impl = wxStringOperations::EncodeChar(ch);
     else
-#endif // wxUSE_UNICODE_UTF8
+#elif wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2
```
I think it could be nice to define `wxUSE_UNICODE_UTF16` for this condition, to avoid repeating it all the time and also to make clear what exactly are we testing for here.

vadz commented on include/wx/string.h:
```diff
@@ -2560,6 +2570,10 @@ class WXDLLIMPEXP_BASE wxString
       if ( !ch.IsAscii() )
           m_impl.assign(wxStringOperations::EncodeNChars(n, ch));
       else
+#elif wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2
```
Sorry, I don't understand the need for this block: isn't the case `ch >= 0x10000` already covered by `!ch.IsAscii()` above?

minoki replied:
No, it isn't, because `!ch.IsAscii()` is under `#if wxUSE_UNICODE_UTF8` block.

Maybe I should change the code from
```c++
#if wxUSE_UNICODE_UTF8
      if ( !ch.IsAscii() )
          m_impl.assign(wxStringOperations::EncodeNChars(n, ch));
      else
#endif
          m_impl.assign(n, (wxStringCharType)ch);
```
to
```c++
#if wxUSE_UNICODE_UTF8
      if ( !ch.IsAscii() )
          m_impl.assign(wxStringOperations::EncodeNChars(n, ch));
      else
          m_impl.assign(n, (wxStringCharType)ch);
#elif wxUSE_UNICODE_UTF16
      if ( !ch.IsBMP() )
          m_impl.assign(wxStringOperations::EncodeNChars(n, ch));
      else
          m_impl.assign(n, (wxStringCharType)ch);
#else
      m_impl.assign(n, (wxStringCharType)ch);
#endif
```
for readability.

vadz replied:
Ah, sorry, I didn't realize it from looking at the diff.

I agree that the logic is more clear in the latter version. I wonder if doing something like

```cpp
#if wxUSE_UNICODE_UTF8 || wxUSE_UNICODE_UTF16
    if ( wxStringOperations::RequiresEncoding(ch) )
        m_impl.assign(wxStringOperations::EncodeNChars(n, ch));
    else
#endif
        m_impl.assign(n, (wxStringCharType)ch);
```
could be even better because it's shorter, but still -- hopefully -- quite clear?

In fact, I even wonder if we could get rid of the `#if`: if `RequiresEncoding()` is just an inlined function returning false in `wxUSE_UNICODE_WCHAR || !wxUSE_UNICODE` case, it and the branch controlled by it should be completely optimized away by the compiler anyhow...

vadz commented on include/wx/string.h:
```diff
@@ -1824,7 +1824,11 @@ class WXDLLIMPEXP_BASE wxString
     else
-#endif // wxUSE_UNICODE_UTF8
+#elif wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2
+    if ( ch >= 0x10000 )
```
Also for readability concerns, could it be worth having an (inline, not to harm performance) `IsUTF16Surrogate()` check instead of using `0x10000` everywhere (notice that in other existing code we also use `<= 0xffff` instead of this test...). I'm actually not sure if it's really going to help, so please feel free to ignore this comment if you think it won't.

vadz commented on include/wx/string.h:
```diff
@@ -3305,6 +3319,10 @@ class WXDLLIMPEXP_BASE wxString
       if ( !ch.IsAscii() )
           m_impl += wxStringOperations::EncodeChar(ch);
       else
+#elif wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2
```
Same here, seems unnecessary?

vadz commented on include/wx/stringops.h:
```diff
@@ -44,9 +44,19 @@ struct WXDLLIMPEXP_BASE wxStringOperationsWchar
+#if wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2
+    struct Utf16CharBuffer
+    {
+        wchar_t data[3];
```
I'd add a default ctor initializing `data` to zeros, you currently do it explicitly in `EncodeChar()`, of course, but it's safer to not allow creating objects with uninitialized data in the first place.

minoki replied:
I followed the definition of `wxUniChar::Utf8CharBuffer` here, but if you prefer, I'll add a default ctor.

vadz replied:
Ah, sorry, I missed this. I don't think consistency between two structs defined in different files is that important and I so I'd still add a default ctor for the new one.

vadz commented on src/common/stringops.cpp:
```diff
@@ -27,6 +27,51 @@
+    }
+    else
+    {
+        buf.data[0] = (wchar_t)(0xD800 | ((v - 0x10000) >> 10));
```
This code is not rocket science, of course, but it still duplicates (in a slightly different form) the existing `{en,de}code_utf16()` functions from `src/common/strconv.cpp`.

Would it be possible to reuse them, e.g. just extract them into inline functions in some private header (existing or new `wx/private/utf16.h`, as you prefer)?

vadz reviewed (commented):
Thanks a lot for fixing this bug! This definitely should be applied anyhow, but if you could please address the few comments above, it would be great.

Thanks again!

minoki committed:
Add wxUSE_UNICODE_UTF16 macro to represent if the internal representation of wxString is UTF-16 or not

The value of this macro is equal to (wxUSE_UNICODE_WCHAR && SIZEOF_WCHAR_T == 2).

minoki committed:
Add some functions to handle supplementary characters

The added functions are:

- wxUniChar::IsBMP()
- wxUniChar::IsSupplementary()
- wxUniChar::HighSurrogate()
- wxUniChar::LowSurrogate()

minoki committed:
Use the added wxUniChar functions in the existing code

minoki commented:
@vadz I updated the branch.

- I defined the macro `wxUSE_UNICODE_UTF16`.
- As for the surrogate pair calculation, I added some functions to `wxUniChar`. They can be called as methods, or static functions taking a `wxUint32`.
    - `wxUniChar::IsBMP()`
    - `wxUniChar::IsSupplementary()`
    - `wxUniChar::HighSurrogate()`
    - `wxUniChar::LowSurrogate()`
    - Existing codes (e.g. `encode_utf16` in src/common/strconv.cpp) are updated to use these functions.
- I added `wxStringOperations::IsSingleCodeUnitCharacter` (negation of `RequiresEncoding` in your suggestion) to eliminate some of the `#if` directives in `wxString` implementation.

minoki commented:
According to [boost.value_init's docs](http://www.boost.org/doc/libs/1_64_0/libs/utility/value_init.htm#compiler_issues), the older versions of C++ Builder have an issue with value-initialization of arrays.

Therefore, if we are going to support them, the constructor of `Utf16CharBuffer` can't just write `: data() {}` to zero-initialize it's content.

Should I care for the older C++ Builder?

vadz commented:
I don't think we care about C++ Builder, so let me just merge this. I had to make another change to `wx/defs.h` to make the check for `SIZEOF_WCHAR_T` work in `wx/chartype.h`, but other than this it looks perfect to me.

Thanks a lot for all your work on this and sorry for the delay with finally merging it!

discnl commented:
This change results in the following repeated warning (with at least MSVC 2008/9 and from searching apparently also at least still with MSVC 2013/12 though I can't verify with other versions right now):

>include\wx/stringops.h(51) : warning C4351: new behavior: elements of array 'wxStringOperationsWchar::SingleCharBuffer::data' will be default initialized

(and the same for line 64 with ANSI compilation)

It looks like the right thing is done since [MSVC 2005/8.0](https://msdn.microsoft.com/en-us/library/1ywe7hcy(v=vs.80).aspx) . However we still support 2002/7.0 as the oldest MSVC version so this should probably be fixed instead of merely suppressing the 'warning'.

vadz commented:
Thanks, I didn't notice this in the AppVeyor logs and MSVS 2015 does _not_ give this warning, so I didn't see it locally neither.

Looking at the code, it doesn't seem like we actually need to initialize these arrays at all, but I guess it's better to still initialize at least the first element, just in case, so I'll commit this as a fix.

vadz commented:
Just FYI: I've applied the same fixes to 3.0 branch as well as they seem to be necessary there too.

Thanks again!

discnl commented:
Thanks for the change, unfortunately now the StringTestCase test doesn't pass (also visible from BuildBot logs, additionally tested with MSVC15 and Debug Win32+x64) :

> .\strings\strings.cpp(1170) : error : Assertion
> Test name: StringTestCase::SupplementaryUniChar
> equality assertion failed
> - Expected: 2
> - Actual  : 12

The Actual value is random.

Failing test case as a oneliner:

>CPPUNIT_ASSERT_EQUAL(2, wxString(wxUniChar(0x23456)).length());

I've verified this only occurs since 4d3aaaf. And initializing the data array completely (only, not partially) fixes it.

It looks like in this case the culprit is `wxStringOperationsWchar::EncodeChar` which writes to at least `data[0]`. This is the newly added code so it might suffice to just add NULL-terminators there (and also at least the other `EncodeChar` variation in the header file) or play it the safe way and init the array fully to 0. I'll leave it up to you.

vadz commented:
Oops, thanks for noticing this.
